Propagate and report the file/line number of registrations on errors.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36536 Functionalize qconcat and register with new registration API
* #36535 Convert qclamp to new operator registration API.
* #36534 Convert qbatch_norm to new operator registration API.
* #36531 Update quantized README for registering operators with new API.
* #36527 Functionalize qadd and register with new registration API.
* #36520 Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops.
* #36513 Add explicit schema for quantized conv/conv_prepack
* #36512 Port detach/detach to new registration API
* #36505 Make it an error to TORCH_LIBRARY the same namespace twice.
* #36390 Port Resize to use new registration API.
* #36389 Update reference to RegisterOperators in error message in Convolution
* **#36264 Propagate and report the file/line number of registrations on errors.**
* #36258 Switch to pybind11 style registration function API.
* #36250 Unconditionally register schema even for manual registration.

Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

ezyang committed:
Propagate and report the file/line number of registrations on errors.

Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/op_registration/op_registration.cpp:
```diff
@@ -140,45 +158,55 @@ Library& Library::_def(c10::either<OperatorName, FunctionSchema>&& name_or_schem
-  if (ns_.has_value()) name.setNamespaceIfNotSet(ns_->c_str());
-  TORCH_CHECK(!(f.dispatch_key_.has_value() && dispatch_key_.has_value()), "Cannot specify a different dispatch key inside a TORCH_LIBRARY_IMPL; please declare a separate TORCH_LIBRARY_IMPL for your dispatch key");
+  if (ns_.has_value()) {
+    TORCH_CHECK(name.setNamespaceIfNotSet(ns_->c_str()), "Attempted to impl ", toString(name), " which is explicitly qualified with a namespace inside a TORCH_LIBRARY, which is not allowed.  If TORCH_LIBRARY's namespace matches the explicitly given namespace, remove the qualifier; otherwise, please place it in an separate TORCH_LIBRARY_IMPL block to make it clear that you are overriding behavior  ...
```
WDYT about wrapping these messages? I guess the idea is to keep them out of the way of the code, but they can be informative and the side scrolling gets pretty crazy. YMMV I guess.

ezyang replied:
At this point it will be kind of annoying to adjust the formatting in the stack diff because of merge conflicts. I could do this after though.

bhosmer reviewed (approved):
😍😍😍

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -115,10 +115,10 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
     listeners_->callOnOperatorRegistered(op);
   } else {
-    checkSchemaCompatibility(op, schema);
+    checkSchemaCompatibility(op, schema, debug);
```
nit: I'd probably also add the std::move here. It doesn't change anything because the `checkSchemaCompatibility` function takes a reference, but that's not immediately visible at the call site, so the call site is easier to check for consistent `std::move` usage if we just add it everywhere where it is valid. That also future-proofs in case `checkSchemaCompatibility` decides to take it by value later.

ezyang replied:
This doesn't make sense to me. Semantically, it doesn't make any sense for schema compatibility to take ownership of the schema. It is after all just going to read the schema, not store it anywhere. Isn't adding a `std::move` here "just because" it's the terminal use obscuring the semantic meaning of the code here?

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -217,7 +217,7 @@ void Dispatcher::cleanup(const OperatorHandle& op, const OperatorName& op_name)
 }
 
-RegistrationHandleRAII Dispatcher::registerFallback(DispatchKey dispatchKey, KernelFunction kernel) {
+RegistrationHandleRAII Dispatcher::registerFallback(DispatchKey dispatchKey, KernelFunction kernel, std::string debug) {
```
you're not using `debug` here, right? In that case, you could also make it a `const&`.

General recommendation: If a function takes an argument by value or by `const&` should not depend on the call site but only on if the function internally takes ownership (e.g. moves the value further) or not. The API contract (i.e. what the call site needs to know) for by-value and `const&` is the same, it's just a compiler optimization. Both calling conventions look the same at the call site, so by reading you cannot distinguish them. That means the call site should ignore if the function takes an argument by value or as `const&` and just `std::move` it in whenever that is valid (and the type is large enough so moving in general makes sense).

ezyang replied:
This one should morally take it by move; the problem is I'm not actually saving debug information for fallbacks (but I should!) There's a more generic problem which should get fixed at some point, which is we're not doing adequate error checking for fallbacks, c.f. this existing comment:

```
RegistrationHandleRAII Dispatcher::registerFallback(DispatchKey dispatchKey, KernelFunction kernel, std::string debug) {
  // TODO: fallbacks clobber each other completely unsafely, unlike regular
  // kernels
```

smessmer commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -102,6 +105,8 @@ class CAFFE2_API OperatorEntry final {
   OperatorName name_;
   c10::optional<FunctionSchema> schema_;
+  c10::optional<std::string> debug_;
+  // INVARIANT: schema_.has_value() == debug_.has_value()
```
should we explicitly model this instead by having a `c10::optional<SchemaInfo>` with

```
struct SchemaInfo final {
    FunctionSchema schema;
    std::string debug;
}
```

That would make the invariant be enforced by the compiler instead of having to rely on the comment here.

ezyang replied:
I considered it, but adding another struct seemed annoying. Up to you.

smessmer commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -320,8 +320,8 @@ struct FunctionSchema {
 
-  void setNamespaceIfNotSet(const char* ns) {
-    name_.setNamespaceIfNotSet(ns);
+  bool setNamespaceIfNotSet(const char* ns) {
```
nit: comment on when this returns true/false. Or even better, make it return an enum with two possible (and descriptive) values.

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang committed:
Update on "Propagate and report the file/line number of registrations on errors."


Pretty simple 1-2 PR:
- Start recording the passed in __FILE__ and __LINE__ from the
  new API in Library, and propagate it to the existing debug
  strings
- Extend OperatorEntry to record a debug string corresponding
  to where the schema came from
- Print out these newly recorded debug strings wherever possible

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929260](https://our.internmc.facebook.com/intern/diff/D20929260)

[ghstack-poisoned]

ezyang commented:
Folded into #36258
